### PR TITLE
Add an upgrade test framework that validates kro upgrades don't break

### PR DESCRIPTION
**Pull request by a-hilaly**
Add upgrade e2e test suite for RGD/instance stability across versions

Validates that upgrading kro from the latest release (currently `v0.8.5`)
to the current build does not break existing RGDs, instances, or managed
resources.

Test flow:
- Install old kro via helm from OCI registry
- Create 17 fixture RGDs covering core features and edge cases
- Run pre-upgrade validation (all RGDs `Active`, instances `ACTIVE`)
- Upgrade to current build via `ko` + `helm template`
- Run post-upgrade validation

Post-upgrade checks:
- All RGDs remain `Active`, conditions `True`, generations bumped
- Controller healthy with zero restarts
- New instances can be created using existing RGDs
- RGD mutation propagates annotations to managed Deployments
- 10 rapid mutations: child resources updated each time, instance stays `ACTIVE`
- Instance deletion garbage collects child resources
- Instance recreation works after deletion
- RGD deletion preserves CRD (`allowCRDDeletion=false`)
- Status projection (`availableReplicas`) works post-upgrade
- `GraphRevision` retention capped at 5 after 10 mutations (GR-aware versions)

Fixture coverage:
- `Deployment`+`Service` with `readyWhen` and status projection
- Conditional resources (`includeWhen`)
- Multi-resource dependency chains
- External references (`externalRef`)
- Validation markers (defaults)
- Nil vs empty: `readyWhen`, `includeWhen`, `forEach`
- Schema defaults: `group` omitted/explicit, `metadata` nil/empty
- Resource ordering stability (reverse alphabetical IDs)

Found bug: legacy `ResourceGraphAccepted` condition from v0.8.x not
pruned on upgrade, left with stale `observedGeneration` (test pending).

Usage: `make test-e2e-upgrade-kind`
       `make test WHAT=upgrade MODE=pre-upgrade|post-upgrade`

### CODE REVIEW

File: scripts/ci/presubmits/e2e-upgrade-tests [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+REPO_ROOT="$(git rev-parse --show-toplevel)"
+cd ${REPO_ROOT}
+
+go build cmd
```
Comment (jakobmoellerdev) [inline scripts/ci/presubmits/e2e-upgrade-tests]: `go build cmd` won't build anything useful — should be `go build ./cmd/...` or `go build ./...`.

File: test/upgrade/gr_retention_test.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+		var grCount int
+		var highestRevision, lowestRevision int64
+		lowestRevision = 999999
+		for _, gr := range grList.Items {
+			if getRGDNameFromGR(gr) != retentionRGDName {
+				continue
+			}
+			grCount++
+			revision, found, _ := unstructured.NestedInt64(gr.Object, "spec", "revision")
+			if found {
```
Comment (jakobmoellerdev) [inline test/upgrade/gr_retention_test.go]: `time.Sleep` is fragile in tests — can both slow CI and still flake. Consider polling with `Eventually` or documenting why it's not sufficient here.

File: test/upgrade/mutation_test.go [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+			// When upgrading from pre-GR version: the initial upgrade created 1 GR
+			// per RGD. Our mutation should create exactly 1 more for the mutated RGD.
+			// When upgrading from GR-aware version: the mutation should create exactly
+			// 1 more than the pre-mutation count.
+			//
+			// In both cases: exactly 1 more GR than before mutation.
+			g.Expect(grCountAfter).To(gomega.Equal(grCountBefore+1),
+				"Expected exactly 1 new GraphRevision after mutation. Before: %d, After: %d, Delta: %d",
+				grCountBefore, grCountAfter, grCountAfter-grCountBefore)
```
Comment (jakobmoellerdev) [inline test/upgrade/mutation_test.go:205]: Same `time.Sleep` concern here.

File: Makefile [hunk 4]
```diff
@@ -380,3 +381,43 @@ deploy-kind: ko deploy-kind-helm ## Deploy kro to a kind cluster
...
+		--set config.allowCRDDeletion=false \
+		--set config.resourceGraphDefinitionConcurrentReconciles=10 \
+		--set config.dynamicControllerConcurrentReconciles=10
+	kubectl wait --for=condition=available --timeout=3m deployment/kro -n kro-system
+	@echo "Old kro $(KRO_UPGRADE_FROM_VERSION) installed"
+
+.PHONY: upgrade-kro
+upgrade-kro: export KO_DOCKER_REPO=kind.local
+upgrade-kro: ko ## Build current kro and upgrade the running deployment (no cluster wipe)
+	make install
```
Comment (jakobmoellerdev) [inline Makefile]: nit: make install should be $(MAKE) install for proper recursive make.
